trax/tf_numpy/extensions/extensions.py

```diff
@@ -453,6 +454,79 @@ def erf(x):
   return tf_np.asarray(tf.math.erf(x.data))
 
 
+def _minus(a, b):
+  return [x for x in a if x not in b]
+
+
+def _compose_output_rep(lhs_rep, rhs_rep, lhs_contraction, rhs_contraction,
+                        lhs_batch, rhs_batch):
+  """ Given lhs representation, rhs representation, contraction and batch dimensions,
+  compose the output representation.
+    e.g., ij, jk, (((1,), (0,)), ((), ())) -> ik
+          aij, ajk, (((2,), (1,)), ((0,), (0,))) -> aik
+  """
+  output_rep = []
+  for dim in lhs_batch:
+    output_rep.append(lhs_rep[dim])
+  for dim in rhs_batch:
+    if rhs_rep[dim] not in output_rep:
+      output_rep.append(rhs_rep[dim])
+
+  for i in _minus(range(len(lhs_rep)), lhs_batch + lhs_contraction):
+    output_rep.append(lhs_rep[i])
+  for i in _minus(range(len(rhs_rep)), rhs_batch + rhs_contraction):
+    output_rep.append(rhs_rep[i])
+  return ''.join(output_rep)
+
+
+def _non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction):
+  """ If it is the general non-batched/single-batched matrix multiplication,
+  use the highly optimized kernel `tf.tensordot` to handle it.
+  """
+  return tf.tensordot(lhs, rhs, axes=(list(lhs_contraction), list(rhs_contraction)))
+
+
+def tf_dot_general(lhs, rhs, dimension_numbers):
+  """ An equivalent general dot operation as that in JAX -
```

Comment:
The docstring format is:
```
"""One-liner summary ended with '.' .

Description (optional).

Args:
  arg1: ...
    ...
  arg2: ...
    ...

Returns:
  ...
"""
```
See `conv` or other existing functions in this file for examples.

Reply:
Sorry about this.